CoThings/Services/NotificationService.swift

```diff
@@ -0,0 +1,95 @@
+//
+//  NotificationService.swift
+//  CoThings
+//
+//  Created by Neso on 2020/05/30.
+//  Copyright © 2020 Rainlab. All rights reserved.
+//
+
+import Foundation
+import Combine
+import UserNotifications
+
+class NotificationService: ObservableObject {
+
+	@Published var showingAlert: Bool = false
+
+	@Published var notifyOnEnter: Bool = UserDefaults.standard.bool(forKey: NotifyOnEnterKey) {
+		didSet {
+			UserDefaults.standard.set(self.notifyOnEnter, forKey: NotifyOnEnterKey)
+			if self.notifyOnEnter {
+				self.notificationRequest()
+			}
+		}
+	}
+
+	@Published var notifyOnExit: Bool = UserDefaults.standard.bool(forKey: NotifyOnExitKey) {
+		didSet {
+			UserDefaults.standard.set(self.notifyOnExit, forKey: NotifyOnExitKey)
+			if self.notifyOnExit {
+				self.notificationRequest()
+			}
+		}
+	}
+
+	@Published var notifyWithSound: Bool = UserDefaults.standard.bool(forKey: NotifyWithSoundKey) {
+		didSet {
+			UserDefaults.standard.set(self.notifyWithSound, forKey: NotifyWithSoundKey)
+			if self.notifyWithSound {
+				self.notificationRequest()
+			}
+		}
+	}
+
+	internal func notificationRequest() {
+		let notificationCenter = UNUserNotificationCenter.current()
+		let options: UNAuthorizationOptions = [.alert, .sound]
+		notificationCenter.requestAuthorization(options: options) {
+			(didAllow, _) in
+			if !didAllow {
+				 DispatchQueue.main.async {
+					self.showingAlert = self.shouldShowAlert()
+				}
+			}
+		}
+		notificationCenter.getNotificationSettings { (settings) in
+			if settings.authorizationStatus != .authorized {
+				 DispatchQueue.main.async {
+					self.showingAlert = self.shouldShowAlert()
+				}
+			}
+		}
+	}
+
+	internal func shouldShowAlert() -> Bool {
+		return notifyOnEnter || notifyOnExit || notifyWithSound
+	}
+
+	func showPushNotificationIfEnabled(for actionEntered: Bool, title: String, message: String) {
```

Comment:
Or instead of `enableChannel`/`disableChannel` we can have
`var enabledChannels: Set<NotificationChannel>()` or even better we can make it immutable and whenever a notification setting changes (like enabling a channel) we can create a new notification settings with enabled channels.

Reply:
you mean we let user have optional notification preference for each channel. this is only for the case of rooms who have beacon. rest of the rooms are manually updated, like washing machines, so we don't need notify them.